File: src/tom/library/strategy/mutraveler/One.java
package tom.library.strategy.mutraveler;
import tom.library.strategy.mutraveler.reflective.AbstractVisitableVisitor;
import jjtraveler.Visitable;
import jjtraveler.reflective.VisitableVisitor;
import jjtraveler.VisitFailure;

/**
 * <code>T(t1,..,ti,..,tN).accept(One(v)) = T(t1,...ti.accept(v),..,tN)</code>
 * if <code>ti</code> is the first child that succeeds.
 * <p>
 * Basic visitor combinator with one visitor argument, that applies
 * this visitor to exactly one child. If no children are visited 
 * successfully, then One(v) fails.
 * <p>
 * Note that side-effects of failing visits to children are not
 * undone.
 */

public class One extends AbstractVisitableVisitor {
  public One(VisitableVisitor v) {
    initSubterm(v);
  }

  public Visitable visit(Visitable any) throws VisitFailure {
    int childCount = any.getChildCount();
    for (int i = 0; i < childCount; i++) {
      try { 
        if(getPosition()!=null) {
          //System.out.println("One.pos = " + getPosition());
          getPosition().down(i);
        }
        Visitable newChild = getArgument(0).visit(any.getChildAt(i));
        if(getPosition()!=null) {
          getPosition().up();
        }
        return any.setChildAt(i,newChild);
      } catch(VisitFailure f) {
        if(getPosition()!=null) {
          getPosition().up();
        }
      }
    }
    throw new VisitFailure();
  }

}

File: tomgo/stable/library/sl/one.go
package sl

import "errors"

// One applies its inner strategy to exactly *one* child of the
// subject — the first one that succeeds. If every child fails (or
// the subject is a leaf), One itself fails.
//
//	One(s)[f(t1,...,ti,...,tn)] = f(t1,...,s[ti],...,tn)
//	                              where ti is the first child s succeeds on.
//	One(s)[c] fails for any leaf c.
type One struct{ abstractCombinator }

func NewOne(s Strategy) Strategy {
	o := &One{}
	o.initSubterm(s)
	return o
}

func (o *One) VisitLight(subject any, intro Introspector) (any, error) {
	count := intro.GetChildCount(subject)
	for i := range count {
		newChild, err := o.args[0].VisitLight(intro.GetChildAt(subject, i), intro)
		if err == nil {
			return intro.SetChildAt(subject, i, newChild), nil
		}
		if !errors.Is(err, ErrVisitFailure) {
			return nil, err
		}
	}
	return nil, ErrVisitFailure
}
